source/hpack.c

```diff
@@ -472,41 +470,44 @@ static const struct aws_http_header *s_get_header_u64(const struct aws_hpack_con
     return aws_hpack_get_header(context, (size_t)index);
 }
 
-/* #TODO pass option whether or not to search for values
- * #TODO I suspect we're not 100% doing the right thing with empty values */
 size_t aws_hpack_find_index(
     const struct aws_hpack_context *context,
     const struct aws_http_header *header,
+    bool search_value,
     bool *found_value) {
 
     *found_value = false;
 
-    /* Check static table */
     struct aws_hash_element *elem = NULL;
-    aws_hash_table_find(&s_static_header_reverse_lookup, header, &elem);
-    if (elem) {
-        *found_value = ((const struct aws_http_header *)elem->key)->value.len;
-        return (size_t)elem->value;
+    if (search_value) {
+        /* Check name-and-value first in static table */
+        aws_hash_table_find(&s_static_header_reverse_lookup, header, &elem);
+        if (elem) {
+            /* If an element was found, check if it has a value */
+            *found_value = ((const struct aws_http_header *)elem->key)->value.len;
+            return (size_t)elem->value;
+        }
+        /* Check name-and-value in dynamic table */
+        aws_hash_table_find(&context->dynamic_table.reverse_lookup, header, &elem);
+        if (elem) {
+            *found_value = ((const struct aws_http_header *)elem->key)->value.len;
+            goto trans_index_from_dynamic_table;
+        }
     }
-    /* If not found, check name only table. Don't set found_value, it will be false */
+    /* Check the name-only table. Note, even if we search for value, when we fail in searching for name-and-value, we
+     * should also check the name only table */
     aws_hash_table_find(&s_static_header_reverse_lookup_name_only, &header->name, &elem);
     if (elem) {
         return (size_t)elem->value;
     }
-
-    /* Check dynamic table */
-    aws_hash_table_find(&context->dynamic_table.reverse_lookup, header, &elem);
+    aws_hash_table_find(&context->dynamic_table.reverse_lookup_name_only, &header->name, &elem);
     if (elem) {
-        /* If an element was found, check if it has a value */
-        *found_value = ((const struct aws_http_header *)elem->key)->value.len;
-    } else {
-        /* If not found, check name only table. Don't set found_value, it will be false */
-        aws_hash_table_find(&context->dynamic_table.reverse_lookup_name_only, &header->name, &elem);
+        goto trans_index_from_dynamic_table;
     }
+    return 0;
 
+trans_index_from_dynamic_table:
     if (elem) {
```

Comment:
[style] we only jump here if `elem` is non-null, replace the if-branch with an `AWS_ASSERT(elem);`

reasoning: it's confusing to have branches that shouldn't be possible based on our programming. there are exceptions for checking results of extremely complex machinery, but any error here would come from within this same function.